zebra-network/src/isolated.rs

```diff
@@ -0,0 +1,78 @@
+//! Code for creating isolated connections to specific peers.
+
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::future::{FutureExt, TryFutureExt};
+
+use tokio::net::TcpStream;
+use tower::{
+    util::{BoxService, Oneshot},
+    Service,
+};
+
+use crate::{peer, BoxedStdError, Config, Request, Response};
+
+/// Use the provided TCP connection to create a Zcash connection completely
+/// isolated from all other node state.
+///
+/// The connection pool returned by `init` should be used for all requests that
+/// don't require isolated state or use of an existing TCP connection. However,
+/// this low-level API is useful for custom network crawlers or Tor connections.
+///
+/// In addition to being completely isolated from all other node state, this
+/// method also aims to be minimally distinguishable from other clients.
+///
+/// Note that this method does not implement any timeout behavior, so callers may
+/// want to layer it with a timeout as appropriate for their application.
+///
+/// # Inputs
+///
+/// - `conn`: an existing TCP connection to use. Passing an existing TCP
+/// connection allows this method to be used with clearnet or Tor transports.
+///
+/// - `user_agent`: a valid BIP14 user-agent, e.g., the empty string.
+pub async fn connect_isolated(
+    conn: TcpStream,
+    user_agent: String,
+) -> Result<BoxService<Request, Response, BoxedStdError>, BoxedStdError> {
+    let handshake = peer::Handshake::builder()
+        .with_config(Config::default())
+        .with_inbound_service(tower::service_fn(|_req| async move {
+            Ok::<Response, BoxedStdError>(Response::Nil)
+        }))
+        .with_user_agent(user_agent)
+        .finish()
+        .expect("provided mandatory builder parameters");
+
+    // We can't get the remote addr from conn, because it might be a tcp
+    // connection through a socks proxy, not directly to the remote. But it
+    // doesn't seem like zcashd cares if we give a bogus one, and Zebra doesn't
+    // touch it at all.
+    let remote_addr = "0.0.0.0:8233".parse().unwrap();
```

Comment:
So this value is bogus? Does it actually have to be a valid value?

Reply:
There's not really any actual specification other than the behaviour of Zebra and Zcashd, and neither of them care about what the value is.  I think that Zcashd might use it for statistics.

Perhaps it would be useful to add a section to #1006 specifying all the values in the handshake and justifying why they're set that way?

Reply:
Yeah that'd be great  thanks!

Reply:
This is the mainnet port, but I don't think that matters.

Let's add a comment saying that we deliberately use the mainnet port on testnet, so someone doesn't try to fix it in future.

Reply:
I'd rather specify the fields of a minimally-distinguished handshake in the RFC.  I don't think that a special comment will be required, because if we test the specified handshake against our normal handshake, we're sure to pick up any change that would break it.

Reply:
There's two kinds of breakage we need to be concerned about here:
* changes that will cause the handshake to fail
* changes that don't cause the handshake to fail, but introduce a stolon version distinguisher

To guard against version distinguishers, we should match the handshake against an exact transcript of the expected handshake. Or check all the fields against the design.

Reply:
Specified the handshake in 7f9c7ff1

Reply:
Added a test for those fields in 11b49856fd26acd9f407b46fcbbe282238a3d5b5